packages/@aws-cdk/core/lib/validation/private/report.ts

```diff
@@ -0,0 +1,222 @@
+        output.push('');
+        output.push('  Occurrences:');
+        for (const construct of constructs) {
+          output.push('');
+          output.push(`    - Construct Path: ${construct.constructPath}`);
+          output.push(`    - Template Path: ${construct.templatePath}`);
+          output.push(`    - Creation Stack:\n\t${construct.constructStack}`);
+          output.push(`    - Resource Name: ${construct.resourceName}`);
+          if (construct.locations) {
+            output.push('    - Locations:');
+            for (const location of construct.locations) {
+              output.push(`      > ${location}`);
+            }
+          }
+        }
+        output.push('');
+        output.push(`  Recommendation: ${plugin.violations[0].recommendation}`);
+        if (plugin.violations[0].fix) {
+          output.push(`  How to fix: ${plugin.violations[0].fix}`);
+        }
+      });
+
+    });
+
+    return output.join(os.EOL);
+  }
+
+  /**
+   * Get the stack trace from the construct node metadata.
+   * The stack trace only gets recorded if the node is a `CfnResource`,
+   * but the stack trace will have entries for all types of parent construct
+   * scopes
+   */
+  private getTraceMetadata(node?: Node): string[] {
+    if (node) {
+      const construct = this.tree.getConstructByPath(node.path);
+      if (construct) {
+        const trace = construct.node.defaultChild?.node.metadata.find(meta => !!meta.trace)?.trace ?? [];
+        return Object.create(trace);
+      }
+    }
+    return [];
+  }
+
+  /**
+   * Construct the stack trace of constructs. This will start with the
+   * resource that has a violation and then go up through it's parents
+   */
+  private getConstructTrace(node: Node, locations?: string[]): ConstructTrace {
+    const trace = this.tree.getTraceFromCache(node.path);
+    if (trace) {
+      return trace;
+    }
+    const metadata = locations ?? this.getTraceMetadata(node);
+    const thisLocation = metadata.shift();
+    const constructTrace = {
+      id: node.id,
+      path: node.path,
+      parent: node.parent ? this.getConstructTrace(node?.parent, metadata) : undefined,
+      library: node.constructInfo?.fqn,
+      libraryVersion: node.constructInfo?.version,
+      location: thisLocation,
+    };
+    this.tree.setTraceCache(constructTrace.path, constructTrace);
+    return constructTrace;
+  }
+
+  /**
+   * - Creation Stack:
+   *     └──  Bucket (validator-test/MyCustomL3Construct/Bucket)
+   *          │ Library: @aws-cdk/aws-s3.Bucket
+   *          │ Library Version: 0.0.0
+   *          │ Location: new Bucket (node_modules/@aws-cdk/aws-s3/lib/bucket.js:688:26)
+   *          └──  MyCustomL3Construct (validator-test/MyCustomL3Construct)
+   *               │ Library: constructs.Construct
+   *               │ Library Version: 10.1.235
+   *               │ Location: new MyCustomL3Construct (/home/packages/@aws-cdk-testing/core-integ/test/integ.core-validations.js:30:9)
```

Comment:
The location is a little strange. For example, this location gives the file/line number that `new Bucket` is called, _not_ where `new MyCustomL3Construct` is called.

Reply:
Yeah you're right. I also think I agree with Rico and we should reverse the tree so that you start from the top and go down. That way you start with what the user is most likely to be familiar with.

Reply:
For example, if the user creates a `pipelines.CodePipeline` they might not know that it creates all these supporting resources for you, like S3 Buckets.

They would create a basic pipeline like this.

```ts
const pipeline = new pipelines.CodePipeline(this, 'Pipeline', {
  codePipeline: codePipeline,
  synth: new pipelines.ShellStep('Synth', {
    input: pipelines.CodePipelineFileSet.fromArtifact(sourceArtifact),
    commands: ['npm ci','npm run build','npx cdk synth'],
  }),
});
```

Then they would see a violation for S3 buckets must be encrypted using KMS keys. But they didn't create any S3 Buckets.